README
This is the README for the Collector plugin.

Collector is usually launched by Nagios, returns its results using SpoolMe or
NSCA (for check results) and StoreMe (for performance data).

There are two command-line parameters:
 * -C: the path to the configuration file (/etc/vigilo-collector/general.conf
   by default)
 * -H: the host name or IP address to check

The general.conf file contains additional configuration parameters, such as the
path to the host configuration files. Those files should be generated by
VigiConf, but an example is given in the provided "host.example" file.

See Collector's internal documentation for more help:
    perldoc ./Collector

The latest documentation is available on the Vigilo website:
    http://www.projet-vigilo.org


debian/changelog
vigilo-collector (3.0.0.g6c79943-1) unstable; urgency=low

  * New upstream release.

 -- François Poirotte <francois.poirotte@c-s.fr>  Sat, 24 Nov 2012 11:39:24 +0100

vigilo-collector (3.0.0.g6c79943-1) unstable; urgency=low

  * New upstream release.

 -- François Poirotte <francois.poirotte@c-s.fr>  Sat, 24 Nov 2012 10:36:37 +0100

vigilo-collector (2.1.2.dev-g0d9fe29-1) unstable; urgency=low

  * New upstream release.

 -- François Poirotte <francois.poirotte@c-s.fr>  Sat, 10 Nov 2012 18:10:39 +0100



lib/sup.pm
################################################################################
## $Id$
##
## sup.pm : PERL function package, Customizable supervision functions for the
##          Collector Nagios Plugin
##
## Copyright (C) 2006-2012 CS-SI
##
## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
################################################################################

package Functions;
use strict;
use warnings;
require Exporter;
use Math::RPN;
use vars qw(@ISA @EXPORT $VERSION);
@ISA=qw(Exporter);
@EXPORT= qw( %Functions );
$VERSION = 1.0;

our %Functions;

$Functions{thresholds_OID_simple} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $OID        = (split('/',$variables->[0]))[1];
    my $warnThresh = $parameters->[0];
    my $critThresh = $parameters->[1];
    my $caption    = $parameters->[2] || "%s";
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $OID") unless exists($response->{$OID});

    my $value      = $response->{$OID};
    return ("UNKNOWN","UNKNOWN: OID $OID not found") unless $Primitive->{"checkOIDVal"}->($value);
    return $Primitive->{"thresholdIt"}->($value,$warnThresh,$critThresh,$caption, $Primitive);
};
$Functions{thresholds_OID_plus_max} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $OID        = (split('/',$variables->[0]))[1];
    my $maxOID     = (split('/',$variables->[1]))[1];
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $OID") unless exists($response->{$OID});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $maxOID") unless exists($response->{$maxOID});

    my $value      = $response->{$OID};
    my $maxValue   = $response->{$maxOID};
    my $warnThresh = $parameters->[0];
    my $critThresh = $parameters->[1];
    my $caption    = $parameters->[2] || "usage: %2.2f%%";

    return ("UNKNOWN","UNKNOWN: OID $OID not found") unless $Primitive->{"checkOIDVal"}->($value);
    return ("UNKNOWN","UNKNOWN: OID $maxOID not found") unless $Primitive->{"checkOIDVal"}->($maxValue);
    return $Primitive->{"thresholdIt"}->($value*100.0/$maxValue,$warnThresh,$critThresh,$caption,$Primitive);
};
$Functions{thresholds_mult} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $OID1       = (split('/',$variables->[0]))[1];
    my $OID2       = (split('/',$variables->[1]))[1];
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $OID1") unless exists($response->{$OID1});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $OID2") unless exists($response->{$OID2});

    my $val1       = $response->{$OID1};
    my $val2       = $response->{$OID2};
    my $warnThresh = $parameters->[0];
    my $critThresh = $parameters->[1];
    my $caption    = $parameters->[2] || "%s";

    return ("UNKNOWN","UNKNOWN: OID $OID1 not found") unless $Primitive->{"checkOIDVal"}->($val1);
    return ("UNKNOWN","UNKNOWN: OID $OID2 not found") unless $Primitive->{"checkOIDVal"}->($val2);
    return $Primitive->{"thresholdIt"}->($val1*$val2,$warnThresh,$critThresh,$caption,$Primitive);
};
$Functions{simple_factor} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $OID        = (split('/',$variables->[0]))[1];
    my $warnThresh = $parameters->[0];
    my $critThresh = $parameters->[1];
    my $factor     = $parameters->[2];
    my $caption    = $parameters->[3] || "%s";

    my $value      = $response->{$OID};
    return ("UNKNOWN","UNKNOWN: OID $OID not found") unless $Primitive->{"checkOIDVal"}->($value);
    # Verifions qu'on a bien un nombre
    return ("UNKNOWN","UNKNOWN: $value") unless $value =~ /^-?(\d+\.?\d*|\.\d+)$/;
    return $Primitive->{"thresholdIt"}->($value*$factor,$warnThresh,$critThresh,$caption, $Primitive);
};
$Functions{table} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $name       = $parameters->[0];
    my $warnThresh = $parameters->[1];
    my $critThresh = $parameters->[2];
    my $caption    = $parameters->[3] || "%s";
    my $OID        = (split('/',$variables->[0]))[1];
    my $descrOID   = (split('/',$variables->[1]))[1];

    # Get the index
    my $index = $Primitive->{"lookup"}->($response,$descrOID,$name);
    return ("UNKNOWN","UNKNOWN: $name not found in $descrOID") if ($index == -1);
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $OID.$index") unless exists($response->{"$OID.$index"});
    my $value=$response->{"$OID.$index"};
    return ("UNKNOWN","UNKNOWN: OID $OID.$index not found") unless $Primitive->{"checkOIDVal"}->($value);
    return $Primitive->{"thresholdIt"}->($value, $warnThresh, $critThresh, $caption, $Primitive);
};
$Functions{table_factor} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $name       = $parameters->[0];
    my $warnThresh = $parameters->[1];
    my $critThresh = $parameters->[2];
    my $factor     = $parameters->[3];
    my $caption    = $parameters->[4] || "%s";
    my $OID        = (split('/',$variables->[0]))[1];
    my $descrOID   = (split('/',$variables->[1]))[1];

    # Get the index
    my $index = $Primitive->{"lookup"}->($response,$descrOID,$name);
    return ("UNKNOWN","UNKNOWN: $name not found in $descrOID") if ($index == -1);
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $OID.$index") unless exists($response->{"$OID.$index"});
    my $value=$response->{"$OID.$index"};
    return ("UNKNOWN","UNKNOWN: OID $OID.$index not found") unless $Primitive->{"checkOIDVal"}->($value);
    return $Primitive->{"thresholdIt"}->($value * $factor, $warnThresh, $critThresh, $caption, $Primitive);
};
$Functions{table_mult} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $name       = $parameters->[0];
    my $warnThresh = $parameters->[1];
    my $critThresh = $parameters->[2];
    my $caption    = $parameters->[3] || "%s";
    my $val1OID    = (split('/',$variables->[0]))[1];
    my $val2OID    = (split('/',$variables->[1]))[1];
    my $descrOID   = (split('/',$variables->[2]))[1];

    # Get the index
    my $index = $Primitive->{"lookup"}->($response,$descrOID,$name);
    return ("UNKNOWN","UNKNOWN: $name not found in $descrOID") if ($index == -1);
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $val1OID.$index") unless exists($response->{"$val1OID.$index"});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $val2OID.$index") unless exists($response->{"$val2OID.$index"});
    my $val1=$response->{"$val1OID.$index"};
    my $val2=$response->{"$val2OID.$index"};
    return ("UNKNOWN","UNKNOWN: OID $val1OID.$index not found") unless $Primitive->{"checkOIDVal"}->($val1);
    return ("UNKNOWN","UNKNOWN: OID $val2OID.$index not found") unless $Primitive->{"checkOIDVal"}->($val2);
    return $Primitive->{"thresholdIt"}->($val1*$val2, $warnThresh, $critThresh, $caption, $Primitive);
};
$Functions{table_used_free} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $name       = $parameters->[0];
    my $warnThresh = $parameters->[1];
    my $critThresh = $parameters->[2];
    my $caption    = $parameters->[3] || "%f%%";
    my $usedOID    = (split('/',$variables->[0]))[1];
    my $freeOID    = (split('/',$variables->[1]))[1];
    my $descrOID   = (split('/',$variables->[2]))[1];

    # Get the index
    my $index = $Primitive->{"lookup"}->($response,$descrOID,$name);
    return ("UNKNOWN","UNKNOWN: $name not found in $descrOID") if ($index == -1);
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $usedOID.$index") unless exists($response->{"$usedOID.$index"});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $freeOID.$index") unless exists($response->{"$freeOID.$index"});
    my $used=$response->{"$usedOID.$index"};
    my $free=$response->{"$freeOID.$index"};
    return ("UNKNOWN","UNKNOWN: OID $usedOID.$index not found") unless $Primitive->{"checkOIDVal"}->($used);
    return ("UNKNOWN","UNKNOWN: OID $freeOID.$index not found") unless $Primitive->{"checkOIDVal"}->($free);
    return $Primitive->{"thresholdIt"}->(($used*100.0)/($free+$used), $warnThresh, $critThresh, $caption, $Primitive);
};
$Functions{table_total_free} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $name         = $parameters->[0];
    my $warnThresh   = $parameters->[1];
    my $critThresh   = $parameters->[2];
    my $caption      = $parameters->[3] || "%f%%";
    my $totalOID     = (split('/',$variables->[0]))[1];
    my $freeOID = (split('/',$variables->[1]))[1];

    # Get the index
    my $total=$response->{"$totalOID"};
    my $free=$response->{"$freeOID"};
    return ("UNKNOWN","UNKNOWN: OID $totalOID not found") unless $Primitive->{"checkOIDVal"}->($total);
    return ("UNKNOWN","UNKNOWN: OID $freeOID not found") unless $Primitive->{"checkOIDVal"}->($free);
    return $Primitive->{"thresholdIt"}->((($total-$free)*100.0)/($total), $warnThresh, $critThresh, $caption, $Primitive);
};
$Functions{table_mult_factor} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $name       = $parameters->[0];
    my $warnThresh = $parameters->[1];
    my $critThresh = $parameters->[2];
    my $factor     = $parameters->[3];
    my $caption    = $parameters->[4] || "%s";
    my $val1OID    = (split('/',$variables->[0]))[1];
    my $val2OID    = (split('/',$variables->[1]))[1];
    my $descrOID   = (split('/',$variables->[2]))[1];

    # Get the index
    my $index = $Primitive->{"lookup"}->($response,$descrOID,$name);
    return ("UNKNOWN","UNKNOWN: $name not found in $descrOID") if ($index == -1);
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $val1OID.$index") unless exists($response->{"$val1OID.$index"});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $val2OID.$index") unless exists($response->{"$val2OID.$index"});
    my $val1=$response->{"$val1OID.$index"};
    my $val2=$response->{"$val2OID.$index"};
    return ("UNKNOWN","UNKNOWN: OID $val1OID.$index not found") unless $Primitive->{"checkOIDVal"}->($val1);
    return ("UNKNOWN","UNKNOWN: OID $val2OID.$index not found") unless $Primitive->{"checkOIDVal"}->($val2);
    return $Primitive->{"thresholdIt"}->($val1 * $val2 * $factor, $warnThresh, $critThresh, $caption, $Primitive);
};
$Functions{sysUpTime} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $criticalTime = $parameters->[0] || 400;
    my $warnTime     = $parameters->[1] || 900;
    my $sysUpTimeOID = (split('/',$variables->[0]))[1];
    my $timestamp    = 0;

    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $sysUpTimeOID") unless exists($response->{$sysUpTimeOID});
    my $sysUpTime    = $response->{$sysUpTimeOID};

    return ("UNKNOWN","UNKNOWN: OID $sysUpTimeOID not found") unless $Primitive->{"checkOIDVal"}->($sysUpTime);
    if ($timestamp = $Primitive->{"date2Time"}->($sysUpTime))
    {
        return $Primitive->{"thresholdIt"}->($timestamp,"\@$warnTime","\@$criticalTime","sysUpTime is $sysUpTime (%d s)",$Primitive);
    }
    return ('UNKNOWN', "UNKNOWN: unable to understand sysUpTime $sysUpTime");
};
$Functions{ifOperStatus} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $interface     = $parameters->[0];
    my $interfaceName = $parameters->[1];
    my $adminWarn     = $parameters->[2] || 'w';
    my $dormantWarn   = $parameters->[3] || 'c';
    my $ifDescrOID    = (split('/',$variables->[0]))[1];
    my $ifAdminStatusOID = (split('/',$variables->[1]))[1];
    my $ifOperStatusOID  = (split('/',$variables->[2]))[1];
    my $ifAliasOID    = (split('/',$variables->[3]))[1];

    my ($state, $answer);
    my $ifIndex = $Primitive->{"lookup"}->($response,$ifDescrOID,$interface);
    return ("UNKNOWN","Interface name ($interface) not found in ifDescr") if ($ifIndex == -1);

    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $ifAdminStatusOID.$ifIndex")
        unless exists($response->{"$ifAdminStatusOID.$ifIndex"});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $ifOperStatusOID.$ifIndex")
        unless exists($response->{"$ifOperStatusOID.$ifIndex"});
    my $ifAdminStatus = $response->{"$ifAdminStatusOID.$ifIndex"};
    my $ifOperStatus  = $response->{"$ifOperStatusOID.$ifIndex"};
    my $ifAlias       = $response->{"$ifAliasOID.$ifIndex"};
    return ("UNKNOWN","UNKNOWN: OID $ifAdminStatusOID.$ifIndex not found") unless $Primitive->{"checkOIDVal"}->($ifAdminStatus);
    return ("UNKNOWN","UNKNOWN: OID $ifOperStatusOID.$ifIndex not found")  unless $Primitive->{"checkOIDVal"}->($ifOperStatus);
    return $Primitive->{"genericIfOperStatus"}->($interfaceName, $ifAdminStatus, $ifOperStatus, $ifAlias, $ifIndex, $adminWarn, $dormantWarn, $Primitive, $debug);
};
$Functions{staticIfOperStatus} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $ifIndex       = $parameters->[0];
    my $interfaceName = $parameters->[1];
    my $adminWarn     = $parameters->[2] || 'w';
    my $dormantWarn   = $parameters->[3] || 'c';
    my $ifDescrOID    = (split('/',$variables->[0]))[1];
    my $ifAdminStatusOID = (split('/',$variables->[1]))[1];
    my $ifOperStatusOID  = (split('/',$variables->[2]))[1];
    my $ifAliasOID    = (split('/',$variables->[3]))[1];

    my ($state, $answer);

    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $ifAdminStatusOID.$ifIndex")
        unless exists($response->{"$ifAdminStatusOID.$ifIndex"});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $ifOperStatusOID.$ifIndex")
        unless exists($response->{"$ifOperStatusOID.$ifIndex"});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $ifDescrOID.$ifIndex")
        unless exists($response->{"$ifDescrOID.$ifIndex"});
    my $ifAdminStatus = $response->{"$ifAdminStatusOID.$ifIndex"};
    my $ifOperStatus  = $response->{"$ifOperStatusOID.$ifIndex"};
    my $ifAlias       = $response->{"$ifAliasOID.$ifIndex"};
    my $ifDescr       = $response->{"$ifDescrOID.$ifIndex"};
    return ("UNKNOWN","UNKNOWN: OID $ifAdminStatusOID.$ifIndex not found") unless $Primitive->{"checkOIDVal"}->($ifAdminStatus);
    return ("UNKNOWN","UNKNOWN: OID $ifOperStatusOID.$ifIndex not found")  unless $Primitive->{"checkOIDVal"}->($ifOperStatus);
    if ($Primitive->{"checkOIDVal"}->($ifDescrOID))
    {
        $interfaceName = "$ifDescr $interfaceName";
    }
    return $Primitive->{"genericIfOperStatus"}->($interfaceName, $ifAdminStatus, $ifOperStatus, $ifAlias, $ifIndex, $adminWarn, $dormantWarn, $Primitive, $debug);
};
$Functions{storage} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $partition         = $parameters->[0];
    my $warnThresh        = $parameters->[1];
    my $critThresh        = $parameters->[2];
    my $percent           = $parameters->[3];
    my $hrStorageDescrOID = (split('/',$variables->[0]))[1];
    my $hrStorageAllocationUnitsOID = (split('/',$variables->[1]))[1];
    my $hrStorageSizeOID  = (split('/',$variables->[2]))[1];
    my $hrStorageUsedOID  = (split('/',$variables->[3]))[1];

    my $hrIndex = $Primitive->{"lookup"}->($response,$hrStorageDescrOID,$partition);
    return ("UNKNOWN","Partition name ($partition) not found in hrStorageDescr") if ($hrIndex == -1);

    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $$hrStorageAllocationUnitsOID.$hrIndex")
        unless exists($response->{"$hrStorageAllocationUnitsOID.$hrIndex"});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $hrStorageSizeOID.$hrIndex")
        unless exists($response->{"$hrStorageSizeOID.$hrIndex"});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $hrStorageUsedOID.$hrIndex")
        unless exists($response->{"$hrStorageUsedOID.$hrIndex"});
    my $hrAU = $response->{"$hrStorageAllocationUnitsOID.$hrIndex"};
    my $hrS  = $response->{"$hrStorageSizeOID.$hrIndex"};
    my $hrU  = $response->{"$hrStorageUsedOID.$hrIndex"};
    return ("UNKNOWN","UNKNOWN: OID $hrStorageAllocationUnitsOID.$hrIndex not found") unless $Primitive->{"checkOIDVal"}->($hrAU);
    return ("UNKNOWN","UNKNOWN: OID $hrStorageUsedOID.$hrIndex not found") unless $Primitive->{"checkOIDVal"}->($hrU);
    return ("UNKNOWN","UNKNOWN: OID $hrStorageSizeOID.$hrIndex not found") unless $Primitive->{"checkOIDVal"}->($hrS);
    my $usedBytes = $hrU*$hrAU;
    my $maxBytes = $hrS*$hrAU;
    return ("UNKNOWN","UNKNOWN: 0 byte Allocation units for storage $partition") if $maxBytes == 0;
    if ($percent)
    {
        my $usagePercentage = $usedBytes*100.0/$maxBytes;
        return $Primitive->{"thresholdIt"}->($usagePercentage,$warnThresh,$critThresh,"Usage: ".sprintf("%.2f",$usedBytes/1024/1024)." MB (%2.2f%%)", $Primitive);
    }
    else
    {
        my $freeBytes = $maxBytes - $usedBytes;
        my $freePercentage = $freeBytes*100.0/$maxBytes;
        $freePercentage = sprintf("%2.2f%%%%",$freePercentage);
        return $Primitive->{"thresholdIt"}->($freeBytes,"@".($warnThresh*1024*1024),"@".($critThresh*1024*1024),"Usage: %d Bytes free ($freePercentage)", $Primitive);
    }
};
$Functions{walk_grep_count} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $pattern    = $parameters->[0];
    my $warnThresh = $parameters->[1];
    my $critThresh = $parameters->[2];
    my $caption    = $parameters->[3] || "%d occurences";
    my $walk       = (split('/',$variables->[0]))[1];

    my $value=0;
    # Get the ifIndex
    foreach my $OID (keys %{$response})
    {
        if ($OID =~ /^$walk\./)
        {
            if ($response->{$OID} =~ /^$pattern\000?$/)
            {
                $value++;
            }
        }
    }
    return $Primitive->{"thresholdIt"}->($value,$warnThresh,$critThresh,$caption,$Primitive);
};
$Functions{statusWithMessage} = sub {
    # NETWORK-APPLIANCE-MIB::miscGlobalStatus.0 = INTEGER: ok(3)
    # NETWORK-APPLIANCE-MIB::miscGlobalStatusMessage.0 = STRING: "The system's global status is normal. "
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;
    my $okValue    = $parameters->[0];
    my $warnValue  = $parameters->[1];
    my $valueOID   = (split('/',$variables->[0]))[1];
    my $msgOID     = (split('/',$variables->[1]))[1];

    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $valueOID") unless exists($response->{$valueOID});
    return ("UNKNOWN", "UNKNOWN: unable to collect value for OID $msgOID") unless exists($response->{$msgOID});
    my $stateValue = $response->{$valueOID};
    my $msgValue   = $response->{$msgOID};
    return ("UNKNOWN","UNKNOWN: OID $valueOID not found") unless $Primitive->{"checkOIDVal"}->($stateValue);
    return ("UNKNOWN","UNKNOWN: OID $msgOID not found") unless $Primitive->{"checkOIDVal"}->($msgValue);
    # TODO: use thresholdIt
    return ("OK","OK: $msgValue") if $stateValue == $okValue;
    return ("WARNING","WARNING: $msgValue") if $stateValue == $warnValue;
    return ("CRITICAL","CRITICAL: $msgValue");
};
$Functions{sup_rpn} = sub {
    my ($parameters, $variables, $response, $debug, $Primitive)=@_;

    my $name       = $parameters->[0];
    my $warnThresh = $parameters->[1];
    my $critThresh = $parameters->[2];
    my $caption    = $parameters->[3];

    my @formula = @$parameters;
    @formula = @formula[4..$#formula];
    # remplacement des OIDs par leur valeur dans la formule
    foreach my $cmpid (0 .. $#formula)
    {
        my $component = $formula[$cmpid];
        if (substr($component, 0, 1) eq ".")
        { # commence par un "." -> c'est un OID
            return ("UNKNOWN","UNKNOWN: $name cannot be computed (can't find $component)") unless exists($response->{$component});
            return ("UNKNOWN","UNKNOWN: $name cannot be computed (wrong value for $component)") unless $Primitive->{"checkOIDVal"}->($response->{$component});
            $formula[$cmpid] = $response->{$component};
        }
    }
    # calcul
    my $result = rpn(@formula);
    return $Primitive->{"thresholdIt"}->($result, $warnThresh, $critThresh, $caption, $Primitive);
};


1;


lib/base.pm
################################################################################
## $Id$
##
## base.pm : PERL function package, Customizable functions for the Collector
##           Nagios Plugin
##
## Copyright (C) 2006-2012 CS-SI
##
## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
################################################################################

package Functions;
use strict;
use warnings;
require Exporter;
use vars qw(@ISA @EXPORT $VERSION);
@ISA=qw(Exporter);
@EXPORT= qw( %Functions %Primitive );
$VERSION = 1.0;

# Hashmap of generic multipurpose routines
our %Primitive = (
"date2Time" => sub {
    # converts a sysuptime string into a number of seconds
    my $date = shift;

    if ($date =~ /^(\d+) days?, (\d+):(\d+):(\d+)\.(\d+)/)
    {
        return (((((($1 * 24) + $2) * 60) + $3) * 60) + $4);
    }
    if ($date =~ /^(\d+) hours?, (\d+):(\d+)\.(\d+)/)
    {
        return (((($1 * 60) + $2) * 60) + $3);
    }
    if ($date =~ /^(\d+) minutes?, (\d+)\.(\d+)/)
    {
        return (($1 * 60) + $2) ;
    }
    if ($date =~ /^(\d+)\.(\d+) seconds?/)
    {
        return ($1);
    }
    return 0;
},
"checkOIDVal" => sub {
    # checks whether the value returned by an SNMP GET or WALKS seems correct
    my $val = shift;
    return 0 unless defined $val;
    return ($val =~ /noSuch(Object|Instance)|endOfMibView|NULL/ ? 0 : 1)
},
"isOutOfBounds" => sub {
    # checks if a value fits is out the "Nagios standard Threshold range"
    # see http://nagiosplug.sourceforge.net/developer-guidelines.html#THRESHOLDFORMAT
    my ($value,$st)=@_;
    my $inside=( ($st =~ /^@/) ? 1 : 0 );
    $st =~ s/^@//;
    $st=":" if ($st eq "");
    if ( $st =~ /:/)
    { # two limits have been given
        my ($low,$upp)=split(/:/,$st);
        return $inside if (! $low && ! $upp);
        return ($inside ? $value <= $upp : $value > $upp) ? 1 : 0 if ($low =~ /^~?$/);
        return ($inside ? $value >= $low : $value < $low) ? 1 : 0 if($upp eq "" );
        return -1 if($low>$upp);
        return( $inside ? ($value >= $low && $value <= $upp) : ($value > $upp || $value < $low) ) ? 1 : 0;
    }
    return ($value >= 0 && $value <= $st) ? 1 : 0 if $inside;
    return ($value < 0 || $value > $st) ? 1 : 0;
},
"thresholdIt" => sub {
    # checks the correctness of the given value, with the warn and crit thresholds
    # see http://nagiosplug.sourceforge.net/developer-guidelines.html#THRESHOLDFORMAT for thresholds syntax
    my ($value,$warnThresh,$critThresh,$caption,$Primitive)=@_;
    $caption=$caption || "%s";
    return("CRITICAL", sprintf("CRITICAL: $caption", $value))   if ($Primitive->{"isOutOfBounds"}->($value, $critThresh));
    return("WARNING",  sprintf("WARNING: $caption",   $value))  if ($Primitive->{"isOutOfBounds"}->($value, $warnThresh));
    return('OK',       sprintf("OK: $caption",       $value));
},
"lookup" => sub {
    # looks for a given pattern in an OID subtree
    # which is supposed to be one-level deep
    # pattern can be a portion of a regexp
    # returns the index of the OID first matching the pattern
    my ($response,$where,$key)=@_;
    foreach my $OID (keys %{$response})
    {
        if ($OID =~ /^$where\./)
        {
            my $content = $response->{$OID};
            if ($content =~ /^0x/){
                # it's looks like an hexadecimal response, let's convert it into BYTE
                $content = substr($response->{$OID}, 2);
                $content =~ s/([a-fA-F0-9][a-fA-F0-9])/chr(hex($1))/eg;
            }
            if ($content =~ /^${key}\000?$/)
            {
                $OID =~ /\.(\d+)$/; # Got it
                return $1;
            }
        }
    }
    return -1;
},
"lookupMultiple" => sub {
    # looks for a given pattern in an OID subtree
    # which is supposed to be one-level deep
    # pattern can be a portion of a regexp
    # returns the list of indexes matching
    my ($response,$where,$key)=@_;
    my @indexes;
    foreach my $OID (keys %{$response})
    {
        if ($OID =~ /^$where\./)
        {
            my $content = $response->{$OID};
            if ($content =~ /^0x/){
                # it's looks like an hexadecimal response, let's convert it into BYTE
                $content = substr($response->{$OID}, 2);
                $content =~ s/([a-fA-F0-9][a-fA-F0-9])/chr(hex($1))/eg;
            }
            if ($content =~ /^${key}\000?$/)
            {
                $OID =~ /\.(\d+)$/; # Got it
                push @indexes,$1;
            }
        }
    }
    return @indexes;
},
"lookupText" => sub {
    # looks for the index betwen an an OID subtree
    # and the numeric representation of the text
    # returns this index
    my ($response,$where,$text)=@_;
    my $numeric = "";
    foreach my $char (split(//,$text))
    {
        $numeric = "$numeric.".ord($char);
    }
    foreach my $OID (keys %{$response})
    {
        if ($OID =~ /^$where\.(\d+$numeric)/)
        {
            return $1;
        }
    }
    return -1;
},
"resultMap" => sub {
    # maps the result given an array of hashmaps describing a regex and a state and message to produce
    # fallback is returned is no match occured
    my ($val,$valFilters,$fallbackState,$fallbackMessage)=@_;
    foreach my $i (@$valFilters)
    {
        if ($val =~ /$$i{"pattern"}/)
        {
            return ($$i{"state"},$$i{"message"});
        }
    }
    return ($fallbackState,$fallbackMessage);
},
"genericHW" => sub {
    my ($response,$descrOID,$stateOID,$okValue,$caption,$debug)=@_;
    my $state="OK";
    my @msg;
    my $nbItems=0;
    foreach my $OID (keys %{$response})
    {
        if ($OID =~ /^$stateOID\.(\d+)/) {
            my $index = $1;
            $nbItems++;
            if ($response->{$OID} =~ /^${okValue}\000?$/) {
                print "ignoring \"".$response->{"$descrOID.$index"}."\" $caption that seems OK\n" if ($debug);
                next;
            }
            push @msg,$response->{"$descrOID.$index"};
            $state="CRITICAL";
        }
    }
    return ("UNKNOWN","UNKNOWN: no $caption found.") if ($nbItems == 0);
    return ("OK","OK: $nbItems $caption(s) OK") if ($state eq "OK");
    return ("CRITICAL","CRITICAL: Failed $caption(s): ".join(',',@msg));
},
"genericIfOperStatus"        => sub {
    my ($interfaceName, $ifAdminStatus, $ifOperStatus, $ifAlias, $ifIndex, $adminWarn, $dormantWarn, $Primitive, $debug)=@_;

    my ($state, $answer);
    my $alias = '';
    if ($Primitive->{"checkOIDVal"}->($ifAlias) && $ifAlias ne '')
    {
        $alias = "($ifAlias)";
    }
    else
    {
        $alias = "(index $ifIndex)";
    }

    $answer = "Interface $interfaceName $alias";
    # if AdminStatus is down - some one made a consious effort to change config
    if ( not ($ifAdminStatus == 1) )
    {
        $answer .= " is administratively down.";
        if ( not defined $adminWarn or $adminWarn eq "w" )
        {
            $state = 'WARNING';
        }
        elsif ( $adminWarn eq "i" )
        {
            $state = 'OK';
        }
        elsif ( $adminWarn eq "c" )
        {
            $state = 'CRITICAL';
        }
        else
        { # If wrong value for authProto, say warning
            $state = 'WARNING';
        }
    } 
    ## Check operational status
    elsif ( $ifOperStatus == 2 )
    {
            $state = 'CRITICAL';
            $answer .= " is down.";
    }
    elsif ( $ifOperStatus == 5 )
    {
        $answer .= " is dormant.";
        if (defined $dormantWarn )
        {
            if ($dormantWarn eq "w")
            {
                $state = 'WARNING';
            }
            elsif($dormantWarn eq "c")
            {
                $state = 'CRITICAL';
            }
            elsif($dormantWarn eq "i")
            {
                $state = 'OK';
            }
        }
        else
        {
            # dormant interface  - but warning/critical/ignore not requested
            $state = 'CRITICAL';
        }
    }
    elsif( $ifOperStatus == 6 )
    {
        $state = 'CRITICAL';
        $answer .= " notPresent - possible hotswap in progress.";
    }
    elsif ( $ifOperStatus == 7 )
    {
        $state = 'CRITICAL';
        $answer .= " down due to lower layer being down.";
    }
    elsif ( $ifOperStatus == 3 || $ifOperStatus == 4  )
    {
        $state = 'CRITICAL';
        $answer .= " down (testing/unknown).";
    }
    else
    {
            $state = 'OK';
            $answer .= " is up.";
    }

    return ($state, "$state: $answer");
},
);

# Initialize sup and metro functions
our %Functions = ();

1;
